File: 6_ Valid Parentheses/solution.py
class Stack:
    def __init__(self):
        self.stack = []

    def push(self, value):
        self.stack.append(value)
    
    def pop(self):
        if not self.is_empty():
            return self.stack.pop()
        else:
            return "Stack is empty"

    def peek(self):
        if not self.is_empty():
            return self.stack[-1]
        else:
            return "Stack is empty"
        
    def is_empty(self):
        return len(self.stack) == 0   
    def size(self):
        return len(self.stack) 


def isValid(s):
    """
    :type s: str
    :rtype: bool
    """
    dictionary = {
        '(':')',
        '[':']',
        '{':'}',
    }
    stack = Stack()
    stack.push(s[0])
    for i in range(len(s)-1):
        if(stack.is_empty()):
            stack.push(s[i+1])
            print("s[i+1] : ",s[i+1])
        elif  stack.peek() in dictionary and dictionary[stack.peek()] == s[i+1] :
            stack.pop()
        else :
            stack.push(s[i+1])

        
    print("sfdvs",stack.peek())

    if stack.is_empty():
        return True
    else :
        return False
    

s="({}}{"

print(isValid(s))

File: 6_ Valid Parentheses/solution.cpp
#include <iostream>
#include <vector>

class Stack{
private:
    std::vector<char> stack; 
public:

    void push(char & str){
        stack.push_back(str);
    }

    char pop(){
        if(!stack.empty()) {
            char str = stack.back();
            stack.pop_back();
            return str;
        } else {
            std::cerr << "Stack empty!" << std::endl;
        }
    }
    char peek(){
        if(!stack.empty()) {
            return stack.back();
        } else {
        std::cerr << "Stack empty!" << std::endl;
        }

    }

    bool isEmpty(){
        return stack.empty();
    }
};

bool isValid(std::string s){
    Stack stack;
    stack.push(s[0]);
    for(int i =1 ; i < s.size() ; i++){
        if (stack.peek() == '(' &&  s[i] == ')'){
            stack.pop();
        } else if (stack.peek() == '{' &&  s[i] == '}'){
            stack.pop();
        } else if (stack.peek() == '[' &&  s[i] == ']'){
            stack.pop();
        } else {
            stack.push(s[i]);
        } 
    }

    if (stack.isEmpty()) {
        return true;
    } else {
        return false;
    }
}

int main(){
    std::string exp = "()[]}{";
    bool result = isValid(exp);
    std::cout << result << std::endl;
    return 0;
}